spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/MergeRowsExec.scala

```diff
@@ -0,0 +1,169 @@
+    val matchedPreds = params.matchedConditions.map(createPredicate(_, joinedAttrs))
+    val matchedProjs = params.matchedOutputs.map(_.map(createProjection(_, joinedAttrs)))
+    val notMatchedPreds = params.notMatchedConditions.map(createPredicate(_, joinedAttrs))
+    val notMatchedProjs = params.notMatchedOutputs.map(_.map(createProjection(_, joinedAttrs)))
+    val projectTargetCols = createProjection(params.targetOutput, joinedAttrs)
+    val nonMatchedPairs = notMatchedPreds zip notMatchedProjs
+    val matchedPairs = matchedPreds zip matchedProjs
+    val rowIdAttrs = params.rowIdAttrs
+    val rowIdProj = createProjection(rowIdAttrs, joinedAttrs)
+    val performCardinalityCheck = params.performCardinalityCheck
+    val emitNotMatchedTargetRows = params.emitNotMatchedTargetRows
+
+    // This method is responsible for processing a input row to emit the resultant row with an
+    // additional column that indicates whether the row is going to be included in the final
+    // output of merge or not.
+    // 1. Found a target row for which there is no corresponding source row (join condition not met)
+    //    - Only project the target columns if we need to output unchanged rows
+    // 2. Found a source row for which there is no corresponding target row (join condition not met)
+    //    - Apply the not matched actions (i.e INSERT actions) if non match conditions are met.
+    // 3. Found a source row for which there is a corresponding target row (join condition met)
+    //    - Apply the matched actions (i.e DELETE or UPDATE actions) if match conditions are met.
+    def processRow(inputRow: InternalRow): InternalRow = {
+      if (emitNotMatchedTargetRows && !isSourceRowPresentPred.eval(inputRow)) {
+        projectTargetCols.apply(inputRow)
+      } else if (!isTargetRowPresentPred.eval(inputRow)) {
+        applyProjection(nonMatchedPairs, inputRow)
+      } else {
+        applyProjection(matchedPairs, inputRow)
+      }
+    }
+
+    var lastMatchedRowId: InternalRow = null
+
+    def processRowWithCardinalityCheck(inputRow: InternalRow): InternalRow = {
+      val isSourceRowPresent = isSourceRowPresentPred.eval(inputRow)
+      val isTargetRowPresent = isTargetRowPresentPred.eval(inputRow)
+
+      if (isSourceRowPresent && isTargetRowPresent) {
+        val currentRowId = rowIdProj.apply(inputRow)
+        if (currentRowId == lastMatchedRowId) {
+          throw new SparkException(
+            "The ON search condition of the MERGE statement matched a single row from " +
+            "the target table with multiple rows of the source table. This could result " +
+            "in the target row being operated on more than once with an update or delete " +
+            "operation and is not allowed.")
+        }
+        lastMatchedRowId = currentRowId.copy()
+      } else {
+        lastMatchedRowId = null
+      }
+
+      if (emitNotMatchedTargetRows && !isSourceRowPresent) {
```

Comment:
Is it expensive to call `processRow` here? I'd prefer not to duplicate it.

Reply:
Well, the way it is written now, we would need to evaluate the is target/source predicate again and will have an extra call for each row.